### PR TITLE
Graphsim

### PR DESCRIPTION
**Pull request by Ericcsr**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/graphsim/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        self.first_frame = torch.from_numpy(self.zipfile['first_frame'])
+
+# Construct fully connected graph
+class MultiBodyGraphCollator:
+    def __init__(self,n_particles):
+        self.n_particles = n_particles
+        self.graph = dgl.from_networkx(nx.complete_graph(self.n_particles))
+
+    def __call__(self,batch):
+        graph_list = [] # Actually we don't need to build a graph each time, for API uniform
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/dataloader.py:56]: What does this mean?

File: examples/pytorch/graphsim/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        data_list = []
+        label_list = []
+        for frame in batch:
+            #graph_list.append(build_dense_graph(n_particles))
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/dataloader.py:60]: Remove comments.

File: examples/pytorch/graphsim/models.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import torch
+import torch.nn as nn
+from torch.nn import functional as F
+import dgl.nn as dglnn
+import dgl.function as fn
+import copy
+from functools import partial
+
+# Assume Global context has been merged to node feature initially
+# Use Relu as default activation
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/models.py:11]: I didn't see how these two lines help explain the class.

File: examples/pytorch/graphsim/models.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+        self.in_layer = InteractionLayer(node_feats-3, # Use velocity only
+                                         node_feats,
+                                         out_node_feats=2,
+                                         out_edge_feats=50,
+                                         edge_fn=edge_fn,
+                                         node_fn=node_fn,
+                                         mode='n_n')
+    
+    # Denormalize Velocity only
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/models.py:97]: Why only on velocity?

File: examples/pytorch/graphsim/models.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    edge_fn : torch.nn.Module
+        Function to update edge feature in message generation
+
+    node_fn : torch.nn.Module
+        Function to update node feature in message aggregation
+
+    mode : str
+        Type of message should the edge carry 
+        nne : [src_feat,dst_feat,edge_feat]
+        n_n : [src_feat-edge_feat]
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/models.py:140]: What are 'nne' and 'n_n' short for?

File: examples/pytorch/graphsim/n_body_sim.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+import argparse
+import os
+import numpy as np
+from math import sin, cos, radians, pi
+import argparse
+
+'''
+This adapted from comes from https://github.com/jsikyoon/Interaction-networks_tensorflow
+Which generate Multi-body Dynamic simulation data used for Interaction network
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/n_body_sim.py:13]: grammer

File: examples/pytorch/graphsim/n_body_sim.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                f_mat[j, i] -= f
+        f_sum[i] = np.sum(f_mat[i], axis=0)
+        acc[i] = f_sum[i]/cur_state[i][0]
+        next_state[i][0] = cur_state[i][0]
+        next_state[i][3:5] = cur_state[i][3:5]+acc[i]*diff_t
+        next_state[i][1:3] = cur_state[i][1:3]+next_state[i][3:5]*diff_t
+    return next_state
+
+# The state is [mass,pos_x,pos_y,vel_x,vel_y]* n_body
+
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/n_body_sim.py:93]: Unify the number of empty lines between functions.

File: examples/pytorch/graphsim/train.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+import dgl
+
+from models import MLP, InteractionNet,PrepareLayer
+from dataloader import MultiBodyGraphCollator, MultiBodyTrainDataset,\
+                       MultiBodyValidDataset, MultiBodyTestDataset
+
+from utils import make_video
+
+# Assuming no initial edge and global feature
+def train(optimizer,loss_fn,model,prep,dataloader,lambda_reg,device):
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/train.py:18]: Space after each parameter.

File: examples/pytorch/graphsim/train.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        label_batch = label_batch.to(device)
+        optimizer.zero_grad()
+        node_feat,edge_feat = prep(graph_batch,data_batch)
+        dummy_relation = torch.zeros(edge_feat.shape[0],1).float().to(device)
+        dummy_global   = torch.zeros(node_feat.shape[0],1).float().to(device)
+        v_pred,out_e = model(graph_batch,node_feat[:,3:5].float(),edge_feat.float(),dummy_global,dummy_relation)
+        loss = loss_fn(v_pred,label_batch)
+        total_loss += float(loss)
+        zero_target = torch.zeros_like(out_e)
+        loss = loss + lambda_reg*loss_fn(out_e,zero_target)*zero_target.numel()/2
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/train.py:32]: What is the second term here?

File: examples/pytorch/graphsim/train.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        for param in model.parameters():
+            reg_loss = reg_loss + lambda_reg*loss_fn(param,torch.zeros_like(param).float().to(device)*param.numel()/2)
+        loss = loss + reg_loss
+        loss.backward()
+        optimizer.step()
+    return total_loss/(i+1)
+
+# One step evaluation
+def eval(loss_fn,model,prep,dataloader,device):
+    total_loss = 0
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/train.py:43]: Prefer to add model.train() in train function, model.eval() in eval function.

File: examples/pytorch/graphsim/dataloader.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        if torch.is_tensor(idx):
+            idx = idx.tolist()
+
+        node_state = self.node_state[idx,:,:]
+        node_label = self.node_label[idx,:,:]
+        return (node_state,node_label)
+
+class MultiBodyTrainDataset(MultiBodyDataset):
+    def __init__(self):
+        super(MultiBodyTrainDataset,self).__init__('./data/n_body_train.npz')
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/dataloader.py:34]: In data generation code, the folder is args, but hard coded here.

File: examples/pytorch/graphsim/utils.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    fig = plt.figure()
+    plt.xlim(-200, 200)
+    plt.ylim(-200, 200)
+    fig_num = len(xy)
+    color = ['ro', 'bo', 'go', 'ko', 'yo', 'mo', 'co']
+    with writer.saving(fig, filename, len(xy)):
+        for i in range(len(xy)):
+            for j in range(len(xy[0])):
+                plt.plot(xy[i, j, 1], xy[i, j, 0], color[j % len(color)])
+            writer.grab_frame()
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/utils.py:26]: Empty line at the end of file. One suggestion is to add such convention in your editor so that it will automatically add that for you.

File: examples/pytorch/tgn/logging.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+Training Interreputed!
```
Comment (sneakerkg) [inline examples/pytorch/tgn/logging.txt:1]: Remove.

File: logging.txt [hunk 14]
```diff
@@ -0,0 +1,5 @@
+Hello World!!
```
Comment (sneakerkg) [inline logging.txt:1]: Remove. One suggestion is to add filter in .ignore in your git setting.

**Review by sneakerkg — CHANGES_REQUESTED** [review]

File: examples/pytorch/graphsim/README.md [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    - Complete Bidirectional Graph
+    - 10 trajectories should be generated
+    - 1000 steps of simulation per trajectory
+
+Dependency
+--------------------------------
+```
+sudo apt install ffmpeg
+pip install opencv-python
+```
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/README.md:25]: Only list the dependency package and version, like:
Dependencies
numpy 1.19.4
pytorch 1.7.1
dgl 0.5.3
networkx 2.5
matplotlib 3.3.4

File: examples/pytorch/graphsim/README.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+One Step Loss Performance, Loss of test data after 40000 training epochs.
+-------------------------
+| Models/Dataset | 6 Body |
+| :-------------- | -----: |
+| Interaction Network in DGL | 80(10) |
+| Interaction Network in Tensorflow | 60 |
+
+-------------------------
+Notice that The datasets are generate dataset directly from Simulator to prevent using Tensorflow to handle the original dataset. The training it very unstable, the even if the minimum loss is achieved from time to time, there are chance that loss will suddenly increase, both in auther's model and out model. The Since the original model hasn't been released, the implementation of this model refers from Tensorflow version implemented in: https://github.com/jsikyoon/Interaction-networks_tensorflow which has consult the first author for some implementation details.
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/README.md:67]: "The training it very unstable" -> "The training is very unstable"
"there are chance that loss will suddenly increase" -> "there are chances that loss will suddenly increase"
"auther's model and out model" -> "author's model and our model"
"The Since the original model hasn't been released" -> "Since the original model hasn't been released"

File: examples/pytorch/graphsim/n_body_sim.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+def get_f(reciever, sender):
+    diff = sender[1:3]-reciever[1:3]
+    distance = norm(diff)
+    if(distance < 1):
+        distance = 1
+    return G*reciever[0]*sender[0]/(distance**3)*diff
+
+# Compute stat according to the paper for normalization
```
Comment (sneakerkg) [inline examples/pytorch/graphsim/n_body_sim.py:65]: Remove extra empty lines.

**Review by sneakerkg — APPROVED** [review]

**Comment by mufeili** [comment]
I think we are missing the entry for "Interaction Networks for Learning about Objects, Relations and Physics" in the subsection of the publication year. Can you help fix that? @Ericcsr @sneakerkg , thanks!
